Utilities/DavixAdaptor/src/DavixFile.cc

```diff
@@ -0,0 +1,346 @@
+}
+
+void DavixFile::create(const std::string &name, bool exclusive /* = false */,
+                       int perms /* = 066 */) {
+  open(name.c_str(), (IOFlags::OpenCreate | IOFlags::OpenWrite | IOFlags::OpenTruncate |
+                      (exclusive ? IOFlags::OpenExclusive : 0)),
+       perms);
+}
+
+void DavixFile::open(const std::string &name, int flags /* = IOFlags::OpenRead */,
+                     int perms /* = 066 */) {
+  open(name.c_str(), flags, perms);
+}
+
+void DavixFile::open(const char *name, int flags /* = IOFlags::OpenRead */, int perms /* = 066 */) {
+  // Actual open
+  if ((name == 0) || (*name == 0)) {
+    edm::Exception ex(edm::errors::FileOpenError);
+    ex << "Cannot open a file without name";
+    ex.addContext("Calling DavixFile::open()");
+    throw ex;
+  }
+
+  if ((flags & IOFlags::OpenRead) == 0) {
+    edm::Exception ex(edm::errors::FileOpenError);
+    ex << "Must open file '" << name << "' at least for read";
+    ex.addContext("Calling DavixFile::open()");
+    throw ex;
+  }
+
+  if (davixPosix && m_fd)
+      close();
+  configureDavixLogLevel();
+  // Translate our flags to system flags
+  int openflags = 0;
+
+  if (flags & IOFlags::OpenRead)
+    openflags |= O_RDONLY;
+
+  DavixError *davixErr;
+  davixReqParams = new RequestParams();
+  // Set up X509 authentication
+  davixReqParams->setClientCertCallbackX509(&X509Authentication, NULL);
+  // Set also CERT_DIR if it is set in envinroment, otherwise use default
+  const char *cert_dir = NULL;
+  if ((cert_dir = getenv("X509_CERT_DIR")) == NULL)
+    cert_dir = "/etc/grid-security/certificates";
+  davixReqParams->addCertificateAuthorityPath(cert_dir);
+
+  davixPosix = new DavPosix(getDavixInstance());
+  m_fd = davixPosix->open(davixReqParams, name, openflags, &davixErr);
+
+  // Check Davix Error
+  if (davixErr) {
+    edm::Exception ex(edm::errors::FileOpenError);
+    ex << "Davix::open(name='" << m_name << "') failed with "
+       << "error '" << davixErr->getErrMsg().c_str() << " and error code " << davixErr->getStatus();
+    ex.addContext("Calling DavixFile::open()");
+    throw ex;
+  }
+  if (m_fd == NULL) {
+    edm::Exception ex(edm::errors::FileOpenError);
+    ex << "Davix::open(name='" << m_name << "') failed as fd is NULL";
+    ex.addContext("Calling DavixFile::open()");
+    throw ex;
+  }
+  m_name = name;
+}
+
+IOSize DavixFile::readv(IOBuffer *into, IOSize buffers) {
+  assert(!buffers || into);
+
+  // Davix does not support 0 buffers;
+  if (buffers == 0)
+    return 0;
+
+  DavixError *davixErr;
+
+  DavIOVecInput input_vector[buffers];
+  DavIOVecOuput output_vector[buffers];
+  IOSize total = 0; // Total requested bytes
+  for (IOSize i = 0; i < buffers; ++i) {
+    input_vector[i].diov_size = into[i].size();
+    input_vector[i].diov_buffer = (char *)into[i].data();
+    total += into[i].size();
+  }
+
+  ssize_t s = davixPosix->preadVec(m_fd, input_vector, output_vector, buffers, &davixErr);
+  if (davixErr) {
```

Comment:
What's the memory management of `davixErr`?  Is this a memory leak?

Use `std::unique_ptr` if we need to own the lifetime of this object.